### PR TITLE
Introduce New Routing Mode Options [API-2240] (2/4)

### PR DESCRIPTION
**Pull request by emreyigit**
The PR introduces the new routing mode options and refactors the old smart mode usage in the code. 

This PR is part of new routing implementations. And built on #899. So, it must be merged after #899

### CODE REVIEW

File: doc/dev/doc/network.md
Comment (gbarnett-hz) [inline doc/dev/doc/network.md:19]: When I read that it sounded like the semantics was deprecated -- this is in reference to `hazelcast.networking.smartRouting` and not the smart routing semantics, that's just been renamed. Maybe clarify it a bit?

File: src/Hazelcast.Net.Tests/Clustering/FailoverTests.cs
Comment (gbarnett-hz) [inline src/Hazelcast.Net.Tests/Clustering/FailoverTests.cs:290]: `smartRouting ? RoutingModes.AllMembers : RoutingModes.SingleMember;` 

I see this logic a few places -- maybe use some helper for it? I understand why you're doing it w.r.t. old logic but new reader without the diff won't know.

**Review by gbarnett-hz — COMMENTED** [review]
LGTM -- I skipped the bits my eyes remembered from #899. @emreyigit could you ping me again after #899 is merged so I can take another look?

File: doc/dev/doc/network.md
Comment (emreyigit) [reply doc/dev/doc/network.md]: I see, you are right. I've paraphrased it. How about now?

File: src/Hazelcast.Net.Tests/Clustering/FailoverTests.cs
Comment (emreyigit) [reply src/Hazelcast.Net.Tests/Clustering/FailoverTests.cs]: I've refactored the smart routing argument to `RoutingModes` directly, no middle transformation anymore.

**Review by gbarnett-hz — DISMISSED** [review]
LGTM -- like #899 I've done on best effort.

File: doc/dev/doc/failover.md
Comment (JamesHazelcast) [inline doc/dev/doc/failover.md:45]:
```suggestion
**(3)** Alternative cluster options. Alternative options are only for connection and authentication. Alternative client options cannot override other fields. General configuration should be done in first client options which is step **1**. For example, here **3**, we cannot change the `~~~~RoutingMode` for each cluster. Also, for server side, provided cluster must have **same partition count**. Client cannot failover to a cluster with different partition count.
```
```suggestion
**(3)** Alternative cluster options. Alternative options are only for connection and authentication. Alternative client options cannot override other fields. General configuration should be done in first client options which is step **1**. For example, here **3**, we cannot change the `RoutingMode` for each cluster. Also, for server side, provided cluster must have **same partition count**. Client cannot failover to a cluster with different partition count.
```
Not sure what purpose the `~~~~` is serving here?

File: doc/dev/doc/network.md
Comment (JamesHazelcast) [inline doc/dev/doc/network.md:29]:
```suggestion
* `@Hazelcast.Networking.RoutingModes.SingleMember` connects to a single member of the cluster (previously Unisocket routing).
* `@Hazelcast.Networking.RoutingModes.MultiMember` connects to subset of members in the cluster.
* `@Hazelcast.Networking.RoutingModes.AllMembers` connects to all members of the cluster (previously Smart routing).
```

File: doc/dev/doc/network.md
Comment (JamesHazelcast) [inline doc/dev/doc/network.md:43]:
```suggestion
Please note that server should support this routing mode which is available with Hazelcast Enterprise 5.5.0.
```

File: doc/dev/doc/network.md
Comment (JamesHazelcast) [inline doc/dev/doc/network.md:46]:
```suggestion
In all members mode, the clients connect to each cluster member. Since each data partition uses the well known and consistent hashing algorithm, each client can send an operation to the relevant cluster member, which increases the overall throughput and efficiency. This mode is the default mode.
```

File: src/Hazelcast.Net/Clustering/ClusterConnections.cs
Comment (JamesHazelcast) [inline src/Hazelcast.Net/Clustering/ClusterConnections.cs:920]: Nit
```suggestion
```

File: src/Hazelcast.Net/Clustering/ClusterState.cs
Comment (JamesHazelcast) [inline src/Hazelcast.Net/Clustering/ClusterState.cs:434]: The Java client has dropped all use of "smart routing" except in the client config option, which is maintained for backwards compatibility only - I'd suggest we do the same for C#

This applies to other areas of the PR too, but I will only mention it here.

File: src/Hazelcast.Net/Exceptions/InvalidPartitionGroupException.cs
Comment (JamesHazelcast) [inline src/Hazelcast.Net/Exceptions/InvalidPartitionGroupException.cs:20]:
```suggestion
    public sealed class InvalidPartitionGroupException : HazelcastException
```
Nit

File: src/Hazelcast.Net/Networking/NetworkingOptions.cs
Comment (JamesHazelcast) [inline src/Hazelcast.Net/Networking/NetworkingOptions.cs:98]: We need to support the "smart routing" config option for the foreseeable future as some of our customers will not be able to update their client configs easily and we don't want to introduce a breaking change.

In the Java client, we do this by overriding `setSmartRouting(boolean)` to set our `RoutingMode` based on the `boolean` provided.

**Review by JamesHazelcast — APPROVED** [review]
While not a C# expert by any means, I've checked over the business logic and it looks solid to me - a few comments to address, but almost all are minor, so approving in advance 👍 Thanks for working on this Emre!

File: src/Hazelcast.Net/Exceptions/InvalidPartitionGroupException.cs
Comment (emreyigit) [reply src/Hazelcast.Net/Exceptions/InvalidPartitionGroupException.cs:20]: Fixed.

File: doc/dev/doc/failover.md
Comment (emreyigit) [reply doc/dev/doc/failover.md:45]: No idea on where it came from. Good catch. Fixed

File: doc/dev/doc/network.md
Comment (emreyigit) [reply doc/dev/doc/network.md]: Fixed.

File: doc/dev/doc/network.md
Comment (emreyigit) [reply doc/dev/doc/network.md]: Fixed.

File: doc/dev/doc/network.md
Comment (emreyigit) [reply doc/dev/doc/network.md]: Fixed.

File: src/Hazelcast.Net/Clustering/ClusterConnections.cs
Comment (emreyigit) [reply src/Hazelcast.Net/Clustering/ClusterConnections.cs:919]: Fixed.

File: src/Hazelcast.Net/Clustering/ClusterState.cs
Comment (emreyigit) [reply src/Hazelcast.Net/Clustering/ClusterState.cs:434]: Fair enough. Refactored that smart routing stays same by setting the `RoutingMode`

File: src/Hazelcast.Net/Networking/NetworkingOptions.cs
Comment (emreyigit) [reply src/Hazelcast.Net/Networking/NetworkingOptions.cs]: As I mentioned above, smart routing stays same by setting the `RoutingMode` as in Java client.

**Review by JamesHazelcast — APPROVED** [review]
Business logic LGTM, thanks for working on this @emreyigit!

**Pull request merged**
